src/txn/txn_ckpt.c

```diff
@@ -894,13 +917,22 @@ __wt_txn_checkpoint(WT_SESSION_IMPL *session, const char *cfg[])
 	 */
 	WT_RET(__wt_session_reset_cursors(session, false));
 
+	/* Ensure the metadata table is open before taking any locks. */
+	WT_RET(__wt_metadata_cursor(session, NULL));
+
 	/*
 	 * Don't highjack the session checkpoint thread for eviction.
 	 *
 	 * Application threads are not generally available for potentially slow
 	 * operations, but checkpoint does enough I/O it may be called upon to
 	 * perform slow operations for the block manager.
+	 *
+	 * Application checkpoints wait until the checkpoint lock is available,
+	 * compaction checkpoints don't.
 	 */
+#define	WT_TXN_SESSION_MASK						\
+	(WT_SESSION_CAN_WAIT | WT_SESSION_NO_EVICTION)
```

Comment:
> The original intent of lock_wait=false was that the operation would not wait for any locks, which is why it used a session flag. We had some exceptions already, but this branch makes a bigger change of only applying lock_wait to the high-level checkpoint / schema locks.
> Can you explain why that is necessary? What problem did the compact change run up against with drops and lock_wait=false?

I'm sorry! (I hate it when people make me figure out why they made a change, my apologies.)

The compact problem wasn't with drop, that was a side-effect of removing the `WT_SESSION_LOCK_NO_WAIT` flag.

Compact wants to do a checkpoint, but not wait on the checkpoint lock. I set `WT_SESSION_LOCK_NO_WAIT` in `__wt_txn_checkpoint()` and that led to a pretty subtle bug.  Checkpoint eventually calls `__sync_file()`, which calls `__wt_evict_file_exclusive_on()`, which calls `WT_WITH_PASS_LOCK()`, which returns `EBUSY` because `WT_SESSION_LOCK_NO_WAIT` is set.

I thought about clearing `WT_SESSION_LOCK_NO_WAIT` after its use in `__wt_txn_checkpoint()`, but you have to do the clear in the lock `op` argument (which gets ugly fast), or in the called function, which separates the flag set from the clear, and kind of defeats the whole idea of a per-session flag.

I also thought about changing `WT_WITH_PASS_LOCK` to call `WT_WITH_LOCK_WAIT` instead of `WT_WITH_LOCK`, but the bug was not only hard to find, it's easy to imagine ways we could re-introduce it in other code paths in the future, that is, it's impossible to guarantee there aren't any problematic lock acquisitions downstream from setting `WT_SESSION_LOCK_NO_WAIT` (and those bugs are potentially workload dependent, so we could fail in the field). To be safe, I think we'd have to change every lock macro other than the top-level ones to ignore the per-session flag, in which case the per-session flag isn't doing anything useful.

The reason I didn't make no-wait versions of all of the macros was because the checkpoint, schema and table locks were the only ones I could find where `WT_SESSION_LOCK_NO_WAIT` looked applicable, and no-wait wasn't needed for the other macros, as far as I could tell. It's certainly possible there are code paths I missed where the per-session flag should reasonably be applied to locks downstream: I looked pretty carefully and didn't see any, but I could easily be wrong.

I could buy the argument there's a downstream lock to which `WT_SESSION_LOCK_NO_WAIT` should apply, and it's easier to pass the flag in the `WT_SESSION_IMPL` structure than drill a hole for a no-wait flag through the API, but as I said, I didn't see any.

In summary, I still prefer this change to a per-session flag, but if you really want the per-session flag, we could put the per-session flag back in and change the non-top-level locking calls to use `WT_WITH_LOCK_WAIT` instead of `WT_WITH_LOCK`.